### PR TITLE
Eliminate the `passableEncoding` hack

### PR DESCRIPTION
**Pull request by FUDCo**
`PassableEncoding` (not to be confused with `encodePassable`) was a non-standard serialization scheme used in some tests, which encoded remote references in-band using Symbols with magic names rather than using the normal marshal package machinery that puts these into the 'slots' element of the standard capdata format.  This bypassed various message filtering and transformation logic in the kernel and also required special methods to be present in the bootstrap vat to translate this encoding and relay messages to their actual intended destinations.

This has now been removed.  The relatively small number of tests which used `passableEncoding` have been updated to use `kmarshal` instead.  Messages and data are now encoded in a form that all the other code understands.  Test messages are also now delivered directly to their destinations without having to count on the existence of a relayer.

In support of this, the controller's `queueToVatRoot` method has been augmented by the addition of a `queueToVatObject` method, allowing tests to send messages to specific objects, targeted using remotable references of the sort returned by `kunser`.  The test support library that a lot of the bootstrap tests use has been updated to use this improved mechanism.

In addition, `kmarshal` itself has been upgraded using a trick that MarkM provided for tagging promises, which allows `kmarshal` to be truly stateless.  The (former) statefulness of `kmarshal` caused problems when the module was imported into different compartments, as each compartment ended up with its own module instance and thus its own version of the state. This in turn caused these compartments to have different beliefs about how particular promises were represented, which caused various things to break.  That's all fixed now.

One wart which has NOT been taken care of in this PR, but which will be addressed in a follow-on PR that we were already planning for, is the duplication of `kmarshal.js` in both the SwingSet package and the liveslots package.  The forthcoming PR will perform a bunch of file renaming and relocation to put a bunch of support tooling, used by both benchmarks and tests, into a package of its own, thereby eliminating a lot of weird dependencies and files in places they don't belong.  As part of this I plan to relocate `kmarshal` into a package of its own that can then be cleanly imported by the kernel, liveslots, and the various tests and test support tooling.

All this is in support of issue #8327

### CODE REVIEW

**Comment by FUDCo** [comment]
@turadg this is the endpoint of the series of commits for you to do your TypeScriptification rebase voodoo on (presumably, once reviewers have signed off on this PR and any review-driven tweaks injected).  The follow-on file and directory refactoring should be very reasonable to do after the TypeScriptification chicken has gone through the GitHub merge snake.

**Comment by FUDCo** [comment]
And, of course, this PR has got the wrong base.  Lemme see if I can tweak it...

File: packages/vats/src/core/lib-boot.js
Comment (warner) [inline packages/vats/src/core/lib-boot.js:203]: Let's see, this removes three utility methods (`messageVat`/`messageVatObject`/`messageVatObjectSendOnly`) from the vat-bootstrap root object, and changes the behavior of `awaitVatObject` to accept (promise for) some record, instead of the `encodePassable` encoding of the same (and does not `encodePassable` the result).

The mainnet chain vat-bootstrap will still have them, and it's `awaitVatObject` will continue to use `decodePassable`/`encodePassable`. I want to think about what that difference means for both tests and mainnet CORE_EVAL proposals that need to do some sort of emergency access.

@dckc What did you have in mind for emergency access (to e.g. the produce/consume PromiseSpace)? It looks like we could have cosmic-swingset code do `controller.queueToVatRoot('bootstrap', 'messageVat', ..)` to make it talk to some other vat (e.g. zoe) and get the results stashed in the `makeEncodePassable()` table, then a `controller.queueToVatRoot('bootstrap', `messageVatObject`, ..)` could target that result and send it a message.

Without them, the cosmic-swingset code would need to figure out the exact `kref` of interest and use the new `controller.queueToVatObject()` to target it.

It's a moot question for mainnet, of course, since the bootstrap vat isn't being upgraded. But developers of emergency fixes will need to take a very different approach depending upon whether they're targeting mainnet (almost certainly) or targeting a new chain whose bootstrap lacks those methods.

@dckc I imagine that emergency fixes are better implemented by a CORE_EVAL which evalutes a whole bundle inside the existing bootstrap vat, where it can make normal `E(produce.whatever).stuff()` calls, and actually `await` the results, and make multiple calls (we should have a name for this fancy interaction, I know let's call it a "program" :-). So using `controller.queueToVatwhatever` from the outside isn't the preferred emergency tool.

I think I've talked myself into being ok with it. And it certainly leaves the codebase looking simpler.

**Review by warner — APPROVED** [review]
Ok I think this is good. Please wait for confirmation from @gibson042 since he was so deeply involved in creating this the first time around, I want to make sure we aren't missing anything.

Also we should let the authors of the modified tests know that `t.deepEqual` on the `kunser` output is no longer the right way to do it (i.e. remind them about the new `matchValue`/etc helpers you added).

File: packages/vats/src/core/lib-boot.js
Comment (FUDCo) [reply packages/vats/src/core/lib-boot.js:203]: > The mainnet chain vat-bootstrap will still have them, and it's awaitVatObject will continue to use decodePassable/encodePassable. I want to think about what that difference means for both tests and mainnet CORE_EVAL proposals that need to do some sort of emergency access.

That method is explicitly marked TEST ONLY and so it frightens me a lot to have it in production.  I don't know that it does anything you'd want to allow on mainnet; I really want to just remove it.  However, when I tried to remove it from `vats/src/core/lib-boot.js` (as I did with all the message relay methods) it turned out there was one test that used the (I think) main chain config instead of the test config that boots with `SwingSet/tools/bootstrap-relay.js`.  I suspect that one test could be fixed and then we could remove the method, but I didn't want to spend the time digging to puzzle it out.

Comment (FUDCo) [reply packages/vats/src/core/lib-boot.js:203]: Also, the relay methods that addressed vats by name (very non-ocap, that) were weird.  Normal `queueToVatRoot` uses the kernel's vat name table, which is statically determined at boot time from the config definition.  The relay methods in `bootstrap-relay` (but **NOT** the ones in `lib-boot`, which also uses the same static lookup that `queueToVatRoot` uses) used a dynamic vat name table, but that table is only updated by vats created via the special test-mode `createVat` bootstrap vat method, not by dynamic vat creations in general, so it was really only suitable for testing anyway.

Comment (dckc) [reply packages/vats/src/core/lib-boot.js:203]: > Let's see, this removes three utility methods (`messageVat`/`messageVatObject`/`messageVatObjectSendOnly`) from the vat-bootstrap root object ...
> @dckc What did you have in mind for emergency access (to e.g. the produce/consume PromiseSpace)?

Only `consumeItem`, `produceItem`, and `resetItem`, arising from #5740

The `//#region testing supports` comment is there to say that `messageVat` and friends were never intended for production use.

> @dckc I imagine that emergency fixes are better implemented by a CORE_EVAL...

Usually, yes. But #5740 imagines adding new devices, and I don't think we can do that with only a CORE_EVAL.

Comment (dckc) [reply packages/vats/src/core/lib-boot.js:203]: Remembering that the code on chain is different seems tricky.

How about a comment with a BIG ASCII ART BANNER to the effect of:

**The agoric-3 mainnet does not run this version of the bootstrap vat.
A mainnet node will show v1:hash-bootstrap in `ps`, indicating
that code from RELEASE/HASH is running.**

Ideally we'd reify the "bootstrap is hard to upgrade" issue in github and cite it there too.

File: packages/boot/test/bootstrapTests/supports.js
Comment (turadg) [inline packages/boot/test/bootstrapTests/supports.js:321]: would be nice to not have to pass in `t`. Is the log output sufficient this way?
```suggestion
export const matchRef = (ref1, ref2, message) =>
  assert(krefOf(ref1) === krefOf(ref2), message);
```

**Review by turadg — COMMENTED** [review]
I haven't looked this over thoroughly but I notice it's based off `benchmarkerator` and not master, and neither are based on master with the TS refactoring yet: https://github.com/Agoric/agoric-sdk/pull/8312#issuecomment-1725790709

File: packages/boot/test/bootstrapTests/supports.js
Comment (FUDCo) [reply packages/boot/test/bootstrapTests/supports.js:321]: `t` is the Ava test object, which you need to run the test predicates else the test failures won't show up in the test report.

**Comment by FUDCo** [comment]
> I haven't looked this over thoroughly but I notice it's based off benchmarkerator and not master, and neither are based on master with the TS refactoring yet

Yes.  Per our discussion the other day I'm counting on you for that refactoring.

File: packages/boot/test/bootstrapTests/supports.js
Comment (turadg) [reply packages/boot/test/bootstrapTests/supports.js:321]: Exception show up, but there's not pretty.
```js
import test from 'ava';

test('t is', t => {
  t.is(1, 2, 'my message');
});

test('assert.equals', t => {
  assert.equal(1, 2, 'my message');
});
```

```
$ ava test/test-utils.js

  ✘ [fail]: t is my message
  ✘ [fail]: assert.equals Error thrown in test
  ─

  t is

  my message

  Difference (- actual, + expected):

  - 1
  + 2

  › packages/internal/test/test-utils.js:5:5



  assert.equals

  Error thrown in test:

  RangeError {
    message: 'my message',
  }

  › packages/internal/test/test-utils.js:9:10

  ─

  2 tests failed
```

I agree this here is better. 

Though I look forward to https://github.com/avajs/ava/issues/1094

File: packages/SwingSet/src/controller/controller.js
Comment (gibson042) [inline packages/SwingSet/src/controller/controller.js]: Remotables can expose methods keyed by any built-in or registered symbol; we should just roll with that here IMO.
```suggestion
      typeof method === 'string' || typeof method === 'symbol' ||
        Fail`method must be a string or symbol, not ${q(typeof method)}`;
```

File: packages/SwingSet/src/lib/kmarshal.js
Comment (gibson042) [inline packages/SwingSet/src/lib/kmarshal.js:54]: This should probably reference https://github.com/endojs/endo/issues/1312 , and maybe also vice versa. We ought to have pseudo-promise support from `makeMarshal` without such gymnastics.

File: packages/SwingSet/src/lib/kmarshal.js
Comment (gibson042) [inline packages/SwingSet/src/lib/kmarshal.js]: This assertion stands out by having a message, unlike the ones I've seen up to now. It might be nice to do the same for at least the preceding `assert.typeof(getter, 'function')`.

File: packages/SwingSet/test/test-promises.js
Comment (gibson042) [inline packages/SwingSet/test/test-promises.js:270]: I'm tempted to suggest refactoring `messageToVat` and `messageToObject` in support of more `E`-like patterns, which is where I think things will ultimately settle and in fact already have in bootstrapTests (e.g., `await EV(watcher).watchLocalPromise('fulfilled', S)`). But that seems out of scope here, and probably needs some more iteration anyway. Bypassing the bootstrap vat for everything except managing dynamic vats is already a big improvement.

File: packages/SwingSet/test/test-promises.js
Comment (gibson042) [inline packages/SwingSet/test/test-promises.js]: I don't think `assert(Array.isArray(restParam))` carries its weight.
```suggestion
  const messageToVat = async (vatName, method, ...args) => {
    const kpid = c.queueToVatRoot(vatName, method, args);
    return awaitRun(kpid);
  };
  const messageToObject = async (presence, method, ...args) => {
    const kpid = c.queueToVatObject(presence, method, args);
    return awaitRun(kpid);
  };
```

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (gibson042) [inline packages/SwingSet/test/upgrade/test-upgrade.js]:
```suggestion
```

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (gibson042) [inline packages/SwingSet/test/upgrade/test-upgrade.js]:
```suggestion
```

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (gibson042) [inline packages/SwingSet/test/upgrade/test-upgrade.js]:
```suggestion
 *   messageToVat: (vatName: string, method: string, ...args: unknown[]) => Promise<unknown>,
 *   messageToVatAndRetain: (vatName: string, method: string, ...args: unknown[]) => Promise<unknown>,
 *   messageToObject: (target: unknown, method: string, ...args: unknown[]) => Promise<unknown>,
 *   messageToObjectAndRetain: (target: unknown, method: string, ...args: unknown[]) => Promise<unknown>,
```

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (gibson042) [inline packages/SwingSet/test/upgrade/test-upgrade.js:166]: _Much_ better.

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (gibson042) [inline packages/SwingSet/test/upgrade/test-upgrade.js:560]: Is this `awaitVatObject` still necessary?
```suggestion
    counters[name] = { presence: counter, kref: krefOf(counter) };
```
which points out that we shouldn't even need the `{ presence, kref }` structure.
```js
  for (const [name, methodName] of Object.entries(counterGetters)) {
    const counter = await messageToObject(exporterRoot, methodName);
    …
    counters[name] = counter;
  }
  …
  const ephCounterRC1 = findRowByKref(rc1, krefOf(counters.ephCounter));
  …
```

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (gibson042) [inline packages/SwingSet/test/upgrade/test-upgrade.js:670]: This kind of call really would benefit from better structure delineation... it takes a lot of cogitation to parse this as an invocation of the "exporter" vat's "syscall-send" method with three arguments, as opposed to e.g. `V('exporter')['syscall-send'](observer, 'acceptImports', [strongObj])`. Although it _is_ a specialized case that could be handled with a local function if things get out of hand.

File: packages/SwingSet/tools/bootstrap-relay.js
Comment (gibson042) [inline packages/SwingSet/tools/bootstrap-relay.js:1]: So much better. Thank you!

File: packages/SwingSet/tools/bootstrap-relay.js
Comment (gibson042) [inline packages/SwingSet/tools/bootstrap-relay.js]: This could probably be made a little more concise.
```js
      // `Far` requires an unfrozen methods object,
      // so we shallow-copy the hardened result from `objectMap`.
      const methods = objectMap(returnValues, makeGetterFunction);
      const remotable = Far(label, { ...methods });
```

File: packages/boot/test/bootstrapTests/supports.js
Comment (gibson042) [inline packages/boot/test/bootstrapTests/supports.js:145]: I think `queueAndRun` would be simpler if it accepted a kpid rather than a callback that returns a kpid.
```suggestion
          {
            const kpid = controller.queueToVatObject(presence, method, args);
            return queueAndRun(kpid);
          },
```
(and probably also a rename in that case, likely related to `awaitRun` in other files)

File: packages/notifier/test/test-publish-kit.js
Comment (gibson042) [inline packages/notifier/test/test-publish-kit.js:275]:
```suggestion
```

File: packages/notifier/test/test-publish-kit.js
Comment (gibson042) [inline packages/notifier/test/test-publish-kit.js:280]:
```suggestion
```
etc.

File: packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js
Comment (gibson042) [inline packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js]:
```suggestion
```

File: packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js
Comment (gibson042) [inline packages/zoe/test/swingsetTests/zoe/test-zoe-upgrade.js]:
```suggestion
```

**Review by gibson042 — APPROVED** [review]
I hope you were able to mechanize some of these extensive changes! But the end result is definitely better, and I appreciate you taking on this cleanup.

File: packages/SwingSet/test/test-promises.js
Comment (FUDCo) [reply packages/SwingSet/test/test-promises.js:270]: Something `E`-like that was syntactically friendlier might be a good idea; it would certainly make the code easier to write and easier on the eyes, though probably out of scope for this PR.  The later PR we have planned for cleaning up the support code and putting it into a proper library of its own might be the place for that.  I'm a bit nervous about `EV` because it leads you down the garden path into thinking you can use it just like you use `E`, which is untrue in important ways.  (Also, I should probably put a big giant warning comment on the definition of `EV` and its associated functions, explaining the important hazard of using them, which possibly even their author(s) might be unaware of.)

Comment (FUDCo) [reply packages/SwingSet/test/test-promises.js:270]: Actually, it strikes me that `messageToObject` and its variants have the same hazard as `EV`, which might well motivate putting all that stuff in one place if only to have a common place to hang the warnings.

File: packages/SwingSet/test/test-promises.js
Comment (FUDCo) [reply packages/SwingSet/test/test-promises.js]: Oh yeah.  I think that's a bit of bleed over from the earlier version of the code, in which `args` was an explicit array parameter rather than using spread.

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (FUDCo) [reply packages/SwingSet/test/upgrade/test-upgrade.js]: Yup.  Change the comments to match the changes in the code...

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (FUDCo) [reply packages/SwingSet/test/upgrade/test-upgrade.js:560]: Perhaps if someone could explain to me what `awaitVatObject` actually _does?_  This was actually somewhat mysterious to me, so I just propagated it through with only superficial changes.  But if we could eliminate it that would also take  care of the issue of having it in the non-test version of the bootstrap vat.

File: packages/SwingSet/tools/bootstrap-relay.js
Comment (FUDCo) [reply packages/SwingSet/tools/bootstrap-relay.js]: Not my code, but that does seem like an improvment.

File: packages/SwingSet/test/upgrade/test-upgrade.js
Comment (FUDCo) [reply packages/SwingSet/test/upgrade/test-upgrade.js:670]: I think something like the `E`/`EV` notation would really help here.

File: packages/boot/test/bootstrapTests/supports.js
Comment (FUDCo) [reply packages/boot/test/bootstrapTests/supports.js:145]: I had the same thought when I first encountered the code this derives from -- it seemed like an obvious simplification.  When I tried it, it blew up in my face.  The purpose of the more complicated pattern is so that the `queueTo...` call happens inside `runThunk`, which waits on the mutex queue before allowing the code that enqueues the message to be executed.  This ensures that the modification of the kernel run queue isn't attempted while a `controller.run()` is in progress.  This is related to the hazard that `EV` introduces, namely that it pretends that message sends and `controller.run()` completions are 1:1 correlated, which _can_ happen and often does, but there's absolutely no guarantee.  There's nothing that couples the completion of `controller.run()` to the settling of the promises that prior `queueTo...` calls returned.  You're just betting that's how it goes; you can set up a test to have that happen, but the onus is on you to arrange for it.  This is very different from being inside a vat, where an `await` or a `.then` can delay indefinitely.

File: packages/SwingSet/src/lib/kmarshal.js
Comment (FUDCo) [reply packages/SwingSet/src/lib/kmarshal.js:54]: I am lead to believe that such support is forthcoming, but not quite as soon as we'd like, so this is what we have for now.

**Pull request merged**
